File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null){
            return ans;
        }
        preorder1(root, ans);
        return ans;
    }
    
    public void preorder1(Node root, List<Integer> ans){
        ans.add(root.val);
        for(Node ch : root.children){
            preorder1(ch, ans);
        }
    }
}

File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */
var preorder = function(root) {
    let ans = [];
    helper(root, ans)
    return ans
};

var helper = function(root, ans){
    if(root==null) return 
    ans.push(root.val)
    for(node of root.children){
        helper(node, ans)
    }
}